flink-runtime/src/test/java/org/apache/flink/runtime/entrypoint/ClusterEntrypointTest.java

```diff
@@ -22,50 +22,279 @@
+                    new TestingHighAvailabilityServicesBuilder().build();
+
+            private ResourceManagerFactory<ResourceID> resourceManagerFactory =
+                    StandaloneResourceManagerFactory.getInstance();
+
+            private Configuration configuration = new Configuration();
+
+            public Builder setHighAvailabilityServices(HighAvailabilityServices haService) {
+                this.haService = haService;
+                return this;
+            }
+
+            public Builder setResourceManagerFactory(
+                    ResourceManagerFactory<ResourceID> resourceManagerFactory) {
+                this.resourceManagerFactory = resourceManagerFactory;
+                return this;
+            }
+
+            public Builder setConfiguration(Configuration configuration) {
+                this.configuration = configuration;
+                return this;
+            }
+
+            public TestingEntryPoint build() {
+                return new TestingEntryPoint(configuration, haService, resourceManagerFactory);
+            }
+        }
+    }
+
+    private static class TestingResourceManagerFactory extends ResourceManagerFactory<ResourceID> {
+
+        private final BiConsumer<ApplicationStatus, String> deregisterAppConsumer;
+
+        private TestingResourceManagerFactory(
+                BiConsumer<ApplicationStatus, String> deregisterAppConsumer) {
+            this.deregisterAppConsumer = deregisterAppConsumer;
+        }
+
+        @Override
+        protected ResourceManager<ResourceID> createResourceManager(
+                Configuration configuration,
+                ResourceID resourceId,
+                RpcService rpcService,
+                HighAvailabilityServices highAvailabilityServices,
+                HeartbeatServices heartbeatServices,
+                FatalErrorHandler fatalErrorHandler,
+                ClusterInformation clusterInformation,
+                @Nullable String webInterfaceUrl,
+                ResourceManagerMetricGroup resourceManagerMetricGroup,
+                ResourceManagerRuntimeServices resourceManagerRuntimeServices,
+                Executor ioExecutor)
+                throws Exception {
+            final SlotManager slotManager =
+                    SlotManagerBuilder.newBuilder()
+                            .setScheduledExecutor(rpcService.getScheduledExecutor())
+                            .build();
+            final JobLeaderIdService jobLeaderIdService =
+                    new TestingJobLeaderIdService.Builder().build();
+            return new TestingResourceManager(
+                    rpcService,
+                    resourceId,
+                    highAvailabilityServices,
+                    heartbeatServices,
+                    slotManager,
+                    NoOpResourceManagerPartitionTracker::get,
+                    jobLeaderIdService,
+                    fatalErrorHandler,
+                    resourceManagerMetricGroup) {
```

Comment:
We don't have to change it now but creating a `TestingResourceManager` is quite a heavy operation. It could be simplified by letting the `DefaultDispatcherResourceManagerComponentFactory` not create a `ResourceManger` but a `ResourceManagerService`. That way we could instantiate a `TestingResourceManagerService` which returns a `TestingResourceManagerGateway` on which we could register the `deregisterAppConsumer`. The `DefaultDispatcherResourceManagerComponentFactory` would then also have to use a `ResourceManagerServiceFactory` instead of a `ResourceManagerFactory`. This could be a follow up issue.